flow/enhancers/socket_info.go

```diff
@@ -28,368 +30,310 @@ import (
 
-			fds := getFDs(process)
-			for _, f := range fds {
-				inode := rFdSocket.FindStringSubmatch(f)
-				if len(inode) == 0 {
-					continue
-				}
-
-				var isUDP = true
-				sinfo, ok := s.cacheUDP.inode.Get(inode[1])
-				if !ok {
-					isUDP = false
-					sinfo, ok = s.cacheTCP.inode.Get(inode[1])
-					if !ok {
-						continue
-					}
-				}
-				si := sinfo.(SocketInfo)
-				parseProcStat(process, &si)
-				if isUDP == true {
-					s.cacheUDP.inode.Set(inode[1], si, cache.DefaultExpiration)
-				} else {
-					s.cacheTCP.inode.Set(inode[1], si, cache.DefaultExpiration)
-				}
-			}
+	switch tcpV6.Type {
+	case tracer.EventConnect, tracer.EventAccept:
+		if socketInfo, err := getProcessInfo(int(tcpV6.Pid)); err == nil {
+			s.addEntry(srcAddr, dstAddr, socketInfo)
 		}
+	case tracer.EventClose:
 	}
 }
 
-func ipToHex(ip46 net.IP) string {
-	s := ""
-	ip := ip46.To4()
-	if ip == nil {
-		ip = ip46
-	}
-	for i := range ip {
-		s += fmt.Sprintf("%02X", ip[(len(ip)-1)-i])
-	}
-	return s
+// LostV6 is called when a tcp v6 event was lost
+func (s *SocketInfoEnhancer) LostV6(uint64) {
 }
 
 func (s *SocketInfoEnhancer) getSocketInfoLocal(f *flow.Flow) bool {
+	var ipPortA, ipPortB string
+	var mappedA, mappedB bool
+	var entry interface{}
+
 	ip := net.ParseIP(f.Network.A)
 	if ip == nil {
 		return false
 	}
 
 	port, _ := strconv.ParseUint(f.Transport.A, 10, 0)
-	ipPortA := fmt.Sprintf("%s:%04X", ipToHex(ip), port)
-	ipPortB := ""
+	ipPortA = hashIPPort(ip, uint16(port))
 
 	ip = net.ParseIP(f.Network.B)
 	if ip != nil {
 		port, _ = strconv.ParseUint(f.Transport.B, 10, 0)
-		ipPortB = fmt.Sprintf("%s:%04X", ipToHex(ip), port)
+		ipPortB = hashIPPort(ip, uint16(port))
 	}
 
-	var cache *socketCache
-	if f.Transport.Protocol == flow.FlowProtocol_UDPPORT {
-		cache = s.cacheUDP
-	} else if f.Transport.Protocol == flow.FlowProtocol_TCPPORT {
-		cache = s.cacheTCP
+	if entry, mappedA = s.local.Get(ipPortA); mappedA {
+		f.SocketA = entry.(*flow.SocketInfo)
 	}
 
-	retried := false
-retry:
-	var siA, siB bool
-	// si.LocalIPPort == ipPortA
-	if local, ok := cache.localAddr.Get(ipPortA); ok {
-		if ca, ok := local.(cacheAddr); ok {
-			if sinfo, ok := cache.inode.Get(ca.inode); ok {
-				if si, ok := sinfo.(SocketInfo); ok {
-					f.SocketA = SocketInfoToFlowSocketInfo(&si)
-					siA = true
-				}
+	// peer (si.LocalIPPort == ipPortB) && (si.RemoteIPPort == ipPortA)
+	if entry, mappedB = s.remote.Get(ipPortB + "/" + ipPortA); mappedB {
+		f.SocketB = entry.(*flow.SocketInfo)
+	}
+
+	return mappedA || mappedB
+}
+
+func (s *SocketInfoEnhancer) scanProc() error {
+	s.local.Flush()
+	s.remote.Flush()
+
+	inodePids := make(map[int]int)
+	namespaces := make(map[uint64]bool)
+	processes := make(map[int]*flow.SocketInfo)
+
+	buildInodePidMap := func() error {
+		// Loop through all fd dirs of process on /proc to compare the inode and
+		// get the pid. Shamelessly taken from github.com/drael/GOnetstat
+		d, err := filepath.Glob("/proc/[0-9]*/fd/[0-9]*")
```

Comment:
I think Glob() is quite slow (could be check with strace) as it do lot's of stat on each file, compare to the os.Open() readdirnames() ...